### PR TITLE
Allocate forms randomly

### PR DESCRIPTION
**Pull request by jon-shipley**

### CODE REVIEW

File: admin/lib/random-generator.js [hunk 1]
```diff
@@ -1,35 +1,50 @@
...
- * @return {string}
- * @throws {Error}
- */
-const getRandom = function (length, chars) {
-  if (!chars) {
-    throw new Error(`Argument 'chars' is undefined`)
-  }
+const random = {
+  /**
+   * @brief Returns a random string
```
Comment (GuyHarwood) [inline admin/lib/random-generator.js]: is brief supported or should it be `@description`?

File: admin/services/data-access/check.data.service.js [hunk 2]
```diff
@@ -333,4 +333,29 @@ checkDataService.sqlUpdate = async function (checkData) {
   return sqlService.update('[check]', checkData)
 }
 
+checkDataService.sqlFindAllFormsUsedByPupils = async function (pupilIds) {
+  const select = `SELECT 
+    f.id,
+    f.name,
+    c.pupil_id     
+  FROM ${sqlService.adminSchema}.${table} c JOIN
```
Comment (GuyHarwood) [inline admin/services/data-access/check.data.service.js]: better to be explicit and specify `INNER JOIN` if that is your intention?

File: admin/spec/service/check-form.service.spec.js [hunk 3]
```diff
@@ -1,92 +1,132 @@
...
-  function setupService (cb) {
-    return proxyquire('../../services/check-form.service', {
-      '../services/data-access/check-form.data.service': {
-        sqlFindActiveForm: jasmine.createSpy().and.callFake(cb),
-        getActiveFormPlain: jasmine.createSpy().and.callFake(cb),
-        findCheckFormByName: jasmine.createSpy().and.callFake(cb),
-        isRowCountValid: jasmine.createSpy().and.callFake(cb)
-      },
-      '../models/check-form': CheckForm
+    it('it should return a check-form', async () => {
```
Comment (GuyHarwood) [inline admin/spec/service/check-form.service.spec.js]: 👍

File: admin/spec/service/check-form.service.spec.js [hunk 4]
```diff
@@ -1,92 +1,132 @@
...
+      spyOn(random, 'getRandomIntInRange').and.throwError('a mock throw')
+      try {
+        await service.allocateCheckForm(availableForms, seenForms)
+      } catch (error) {
+        expect(error.message).toBe('Error allocating checkForm: a mock throw')
+      }
+    })
+  })
+
+  describe('#prepareQuestionData()', () => {
```
Comment (GuyHarwood) [inline admin/spec/service/check-form.service.spec.js]: tests looking much more readable 💯

File: admin/services/data-access/check.data.service.js (hunk 2 above)
Comment (jon-shipley) [reply admin/services/data-access/check.data.service.js]: `Join` defaults to `INNER` on MS-SQL but no problem being explicit about it.  Pushed.

**Pull request merged**